circle/src/util.rs

```diff
@@ -0,0 +1,177 @@
+use p3_field::extension::Complex;
+#[cfg(test)]
+use p3_field::extension::ComplexExtendable;
+use p3_field::{ExtensionField, Field};
+#[cfg(test)]
+use p3_matrix::dense::RowMajorMatrix;
+#[cfg(test)]
+use p3_util::{log2_strict_usize, reverse_slice_index_bits};
+
+/// Get the cfft polynomial basis.
+/// The basis consists off all multi-linear products of: y, x, 2x^2 - 1, 2(2x^2 - 1)^2 - 1, ...
+/// The ordering of these basis elements is the bit reversal of the sequence: 1, y, x, xy, (2x^2 - 1), (2x^2 - 1)y, ...
+/// We also need to throw in a couple of negative signs for technical reasons.
+#[cfg(test)]
+pub(crate) fn circle_basis<F: ComplexExtendable>(point: Complex<F>, log_n: usize) -> Vec<F> {
+    if log_n == 0 {
+        return vec![F::one()];
+    }
+
+    let mut basis = vec![F::one()];
+    basis.reserve(1 << log_n);
+
+    // First compute the repeated applications of the squaring map π(x) = 2x^2 - 1
+    let mut cur = point.real();
+    for _ in 0..(log_n - 1) {
+        for i in 0..basis.len() {
+            basis.push(basis[i] * cur);
+        }
+        cur = F::two() * cur.square() - F::one();
+    }
+
+    // Bit reverse, and compute the second half of the array,
+    // which is just each element of the first half times y
+    reverse_slice_index_bits(&mut basis);
+    for i in 0..basis.len() {
+        basis.push(basis[i] * point.imag());
+    }
+
+    // Negate each element each time the binary representation of its index has a pair of adjacent ones,
+    // or equivalently, if the number of adjacent ones is odd.
+    // This comes from a peculiarity in how we compute the CFFT:
+    // The butterfly zips the first half of the domain with the second half reversed, because that maps each point
+    // to its involution. After each layer, the second half is still in reverse order, so we should use the twiddles
+    // in reverse order as well, but we ignore that and use the same twiddles for both halves.
+    // Using t(g^(N-k)) instead of t(g^k) just adds a negative sign. It turns out the number of negations is the number
+    // of adjacent ones in the index.
```

Comment:
I would add this comment also to the CFFT code, pointing out the difference to the paper.